builtin/credential/aws/path_login.go

```diff
@@ -888,6 +1048,389 @@ func (b *backend) pathLoginRenew(
+	}
+
+	endpoint := "https://sts.amazonaws.com"
+
+	if config != nil {
+		if config.IAMServerIdHeaderValue != "" {
+			err = validateVaultHeaderValue(headers, parsedUrl, config.IAMServerIdHeaderValue)
+			if err != nil {
+				return logical.ErrorResponse(fmt.Sprintf("error validating %s header: %s", iamServerIdHeader, err)), nil
+			}
+		}
+		if config.STSEndpoint != "" {
+			endpoint = config.Endpoint
+		}
+	}
+
+	clientArn, accountID, err := submitCallerIdentityRequest(method, endpoint, parsedUrl, body, headers)
+	if err != nil {
+		return logical.ErrorResponse(fmt.Sprintf("error making upstream request: %s", err)), nil
+	}
+	canonicalArn, principalName, sessionName, err := parseIamArn(clientArn)
+	if err != nil {
+		return logical.ErrorResponse(fmt.Sprintf("Error parsing arn: %s", err)), nil
+	}
+
+	roleName := data.Get("role").(string)
+	if roleName == "" {
+		roleName = principalName
+	}
+
+	roleEntry, err := b.lockedAWSRole(req.Storage, roleName)
+	if err != nil {
+		return nil, err
+	}
+	if roleEntry == nil {
+		return logical.ErrorResponse(fmt.Sprintf("entry for role %s not found", roleName)), nil
+	}
+
+	if !strutil.StrListContains(roleEntry.AllowedAuthTypes, iamAuthType) {
+		return logical.ErrorResponse(fmt.Sprintf("auth method iam not allowed for role %s", roleName)), nil
+	}
+
+	// The role creation should ensure that either we're inferring this is an EC2 instance
+	// or that we're binding an ARN
+	if roleEntry.BoundIamPrincipalARN != "" && roleEntry.BoundIamPrincipalARN != canonicalArn {
+		return logical.ErrorResponse(fmt.Sprintf("IAM Principal %q does not belong to the role %q", clientArn, roleName)), nil
+	}
+
+	policies := roleEntry.Policies
+	rTagMaxTTL := time.Duration(0)
+
+	inferredEntityType := ""
+	inferredEntityId := ""
+	if roleEntry.RoleInferredType == ec2EntityType {
+		instance, err := b.validateInstance(req.Storage, sessionName, roleEntry.InferredAWSRegion, accountID)
+		if err != nil {
+			return logical.ErrorResponse(fmt.Sprintf("failed to verify %s as a valid EC2 instance in region %s", sessionName, roleEntry.InferredAWSRegion)), nil
+		}
+
+		// build a fake identity doc to pass on metadata about the instance to verifyInstanceMeetsRoleRequirements
+		identityDoc := &identityDocument{
+			Tags:        nil, // Don't really need the tags, so not doing the work of converting them from Instance.Tags to identityDocument.Tags
+			InstanceID:  *instance.InstanceId,
+			AmiID:       *instance.ImageId,
+			AccountID:   accountID,
+			Region:      roleEntry.InferredAWSRegion,
+			PendingTime: instance.LaunchTime.Format(time.RFC3339),
+		}
+
+		roleTagResp, validationError, err := b.verifyInstanceMeetsRoleRequirements(req.Storage, instance, roleEntry, roleName, identityDoc)
+		if err != nil {
+			return nil, err
+		}
+		if validationError != nil {
+			return logical.ErrorResponse(fmt.Sprintf("Error validating instance: %s", validationError)), nil
+		}
+
+		if roleTagResp != nil {
```

Comment:
@joelthompson 

> Just for my own sanity, is the crux of the issue here that role tags contain fields specific to both instances and the auth type used by instances?

The issue is that role tags are supposed to represent a strict subset of the capabilities on the role, and not partial. Hence, we can't choose to keep certain things out, like, `disallow_reauthentication` and `allow_instance_migration`.

Regarding the migration path, since we are thinking about making a role represent a single auth type, users would need to create a different role to authenticate with `iam` auth type anyways. The role tags created with previous role would only be applicable to that role. Clients won't be able to assume a broader set of policies as the role would remain the same.

I do agree with you that having role tags would reduce role sprawl. However, the right way IMO to do that is to have new role tags that are specific to `iam` auth types. There is a version in role tags. You are up for it, you could write version 2 of role tags that contains auth type and fields specific to auth types.

> My concern here is about people changing allowed auth types on the role. Do we just make that immutable on the role?

Yes! We should make it immutable. That way, there is no interference between the two auth types.

> We could just set an auth parameter of auth_type and ask customers ;) On an actual note, I think we should just error out in that scenario.

If the auth type on the role is immutable, we won't run into this problem at all.